Library/Util.ts

```diff
@@ -146,33 +146,67 @@ class Util {
     }
 
     /**
-     * Validate that an object is of type { [key: string]: string }
+     * Using JSON.stringify, by default Errors do not serialize to something useful:
+     * Simplify a generic Node Error into a simpler map for customDimensions
+     * Custom errors can still implement toJSON to override this functionality
      */
-    public static validateStringMap(obj: any): { [key: string]: string } {
-        var map: { [key: string]: string };
-        if(typeof obj === "object") {
-            map = <{ [key: string]: string }>{};
-            for (var field in obj) {
-                var property = obj[field];
-                var propertyType = typeof property;
-                if (propertyType !== "string") {
-                    if (property != null && typeof property.toString === "function") {
-                        property = property.toString();
-                    } else {
-                        Logging.info("key: " + field + ", invalid property type: " + propertyType);
-                        continue;
-                    }
-                }
+    protected static extractError(err: Error): { message: string, code: string } {
+        // Error is often subclassed so may have code OR id properties:
+        // https://nodejs.org/api/errors.html#errors_error_code
+        const looseError = err as any;
+        return {
+            message: err.message,
+            code: looseError.code || looseError.id || "",
+        }
+    }
 
-                map[field] = property.trim(0, Util.MAX_PROPERTY_LENGTH);
-            }
-        } else {
+    /**
+     * Capture all properties on an object rather than relying on JSON.stringify to do this,
+     * which may yield an unwanted result, depending on property visibility
+     */
+    protected static extractObject(origProperty: any): { [key: string]: any } {
+        if (origProperty instanceof Error) {
+            return Util.extractError(origProperty);
+        }
+        return origProperty;
+    }
+
+    /**
+     * Validate that an object is of type { [key: string]: any }
+     */
+    public static validateStringMap(obj: any): { [key: string]: string } {
+        if(typeof obj !== "object") {
             Logging.info("Invalid properties dropped from payload");
+            return;
         }
+        const map: { [key: string]: string } = {};
+        for (let field in obj) {
+            let property: string = '';
+            const origProperty: any = obj[field];
+            const propType = typeof origProperty;
+
+            if (propType === "string" || propType === "number" || propType === "boolean") {
+                property = origProperty.toString();
+            } else if (origProperty === null || propType === "undefined") {
+                property = "";
+            } else if (propType === "object" && typeof origProperty.toJSON === "function") {
+                property = origProperty.toJSON();
```

Comment:
Stringify should still be called on this, as `toJSON` should return an object for stringifying, not an already serialized representation. As such, we can't guarantee its return value is a string until it's stringified. (For reference, see this comment in Crockford's implementation of JSON.stringify https://github.com/douglascrockford/JSON-js/blob/master/json2.js#L25-L32)